torch/utils/_python_dispatch.py

```diff
@@ -1,34 +1,51 @@
-import torch
 import contextlib
 from typing import Iterator
 
-# Context manager that causes all pytorch operators to dispatch to the passed-in
-# type's __torch_dispatch__ function.
-# operation that accepts no tensors but returns a tensor.
-#
-# enable_python_mode is affected by torch._C._DisableTorchDispatch.
-#
+from torch.utils._mode_utils import _enable_mode, _ModeInfo
+
 # NB: Calling an operator inside __torch_dispatch__ does go through
 # __torch_dispatch__ again. Please use _DisableTorchDispatch inside
 # __torch_dispatch__ to prevent infinite recursion.
 #
 # TODO: Limitations and things about enable_python_mode we should fix before exposing it:
-# - it currently cannot be nested. This should be simple to implement; we need a
-#   stack of TorchDispatchTypeObjects and the next bullet point.
-# - We need a better user-facing api for torch._C._DisableTorchDispatch that
+# - We need a better user-facing api for _DisableTorchDispatch that
 #   is able to selectively disable __torch_dispatch__ of a particular class.
 # - It doesn't work with the tensor constructors (torch.tensor, torch.Tensor)
 # - Better name (see https://github.com/pytorch/pytorch/pull/63496#discussion_r694091694)
 @contextlib.contextmanager
-def enable_python_mode(cls) -> Iterator[None]:
-    if cls.__torch_dispatch__ is torch.Tensor.__torch_dispatch__:
-        raise ValueError('The class passed to enable_python_mode '
-                         'must have a non-default __torch_dispatch__ classmethod')
-    if not isinstance(cls, type) or not issubclass(cls, (torch.Tensor,)):
-        raise ValueError('The argument passed to enable_python_mode '
-                         'must be the type of a Tensor subclass')
-    torch._C._enter_python_mode(cls)
-    try:
-        yield
-    finally:
-        torch._C._exit_python_mode()
+def enable_python_mode(mode, *, replace=None, ignore_preexisting=False) -> Iterator[None]:
+    """
+    Context manager that causes all pytorch operators to dispatch to the passed-in
+    type's __torch_dispatch__ function, including operations that accepts no tensors
+    but returns a tensor.
+
+    This function is non-compositional; if there is already an existing mode,
+    it will raise an error; prefer using :func:`push_python_mode` if your
+    ``__torch_dispatch__`` implementation can defer to an inner mode.
+
+    This function is safe to use inside a ``__torch_dispatch__`` mode handler,
+    as the mode is guaranteed to be disabled in this context.  You can use
+    this context manager to reinstate the mode so that calls to overridable
+    APIs recursively call back into your mode handler (this can easily cause
+    infinite loops, so use with care!)
+
+    enable_python_mode is affected by _DisableTorchDispatch.
```

Comment:
Lol _DisableTorchDispatch is not public API, but I guess that's OK, we haven't released enable_python_mode as public API yet either

Reply:
BTW, this is not symmetric with how DisableTorchFunction interacts with torch function mode. Disabling torch function means we no longer try to dispatch off of arguments, but a mode may still be valid and this situation could occur if the mode had originally returned NotImplemented for tensor subclass arguments

Reply:
> BTW, this is not symmetric with how DisableTorchFunction interacts with torch function mode. Disabling torch function means we no longer try to dispatch off of arguments, but a mode may still be valid and this situation could occur if the mode had originally returned NotImplemented for tensor subclass arguments

Not sure if we want to unify on one behavior, but I think had some cases in AOTAutograd tracing where it was important that the mode also disabled the mode behavior. I'll try to isolate the cases to better explain

Reply:
Will be very interested to see it. My default assumption is that AOTAutograd was relying on the behavior in a bad way and writing the implementation slightly different would resolve it. If you can't isolate it I'm also happy to look into the complicated situation.